### ISSUE TITLE
Dataset Syncronization

### ISSUE THREAD

**Issue by chavinlo** [chavinlo/distributed-diffusion#12]
Check https://github.com/learning-at-home/dalle-hivemind

1.- It creates a Trainer, with argument "train_dataset" pointing to the function training_dataset from the task import
2.- Said function checks if the dataset already exists, if not it calls another function "make_dataset"
3.- Loads the dataset in this way:
-  HuggingFace's dataset loader set to streaming mode
- it randomly selects a certain amount of images based on the provided seed (shuffle_seed)
- then converts it to torch format and returns it

Divide the dataset in 4 chunks, lets say 25 images each, 100 images total.
4 peers: Peer 1 loads chunk 1, Peer 2 chunk 2, and so on.
Once ALL peers finish training, and syncronize, would that count as an entire epoch through the entire dataset?

**Comment by chavinlo** [comment]
Although this has already been implemented, I think a better solution is to still follow part of the Hivemind's dalle methodology:

Rather than delivering them in chunks, the trainer should retrieve them when it is needed.
Have a box of 100 images, all of which are retrieved from a S3 bucket (or wathever)
Then the trainer grabs one of those images and trains on it
while this happens, the box must grab another image from the internet and have it ready to serve it to the trainer
Then the trainer grabs another one

something like that, and although hivemind's did not implemented any dataset tracking, I think it would be better if we do here for more efficient training, so the dataset server is gonna stay at least for tracking and statistics, and in order to manage what images are being used, the ds server communicates with the trainer.
